README.md
`deepcolour` is a stateful library for the parsing, processing and formatting of RGBA colours, as well as vector spaces.

# Features

* All Colour values **unclamped** and normalized to `0-1`.
* Most operations are **in-place**.
* Stateful **HSV** properties.  
* Conversion to and from **CSS**.

# Example

``` javascript
const Colour = require('deepcolour')

let colour = new Colour(1,0,0,1)

console.log( colour.hex )
// #ff0000

// make colour cyan by shifting hue
colour.hue = 0.5

console.log( colour.hex )
// #00ffff

// Set the colour with a string
colour.set('white')

// Retrieve value and saturation
console.log( colour.value, colour.saturation )
// 1,0
```

# API

## Number Properties

* **red** : RGB red value.
* **green** : RGB green value.
* **blue** : RGB blue value.
* **alpha** : Alpha value.
* **hue** : HSV hue. The range of the HSV colour circle is `0-1`. See [Hue Preservation](#hue-preservation).
* **saturation** : HSV saturation.
* **value** : HSV value.

## String Properties

* **hex** : CSS hexadecimal color. `#112233`
* **css** : CSS color. If the colour can be represented as a hexadecimal color, `hex` will be returned. Otherwise a CSS `rgb()` or `rgba()` will be returned.

## Set Methods

All methods beginning with set may be chained together.

**set** *( anything )* Safely calls other set function based on type.

**setDefault** *()* Sets the colour to opaque black, `[0,0,0,1]`.

**setArguments** *( arguments )* Sets from an arguments object. The following formats are supported:
* `( [ red [, green [, blue, [, alpha ] ] ] )`
* `( 'string' [, alpha ] )`

**set8BitArray** *( Array-like )*
Set from an array of numbers in the range 0-255. Up to four values will be read, in the order RGBA. This method will also work for slices of `Buffer` objects, allowing it to work with image libraries such as [png-js](https://www.npmjs.com/package/pngjs). *Throws error on invalid input.*

**setRGB** *( Number, Number, Number )* Sets red, green and blue values. If non-numeric values are passed, existing values are preserved.

**setHSV** *( Number, Number, Number )* Sets hue, saturation and value values. If non-numeric values are passed, existing values are preserved.

**setKeys** *( object )* Sets properties from an object. The following keys are supported: `['red','r','green','g','blue','b','alpha','a','hue','h','saturation','sat','s','value','v','hex','css']`

**setAlpha** *( Number )* Set alpha value.

**setRandom** *()* Set red, green and blue to random values in standard range.

**setChannel** *( index, Number)* Set the channel `index` to a value.

**setChannelHex** *( index, string )* Set the channel `index` to a hexadecimal value. The string may be 1 or 2 digits.

## Output Methods

**toString** *()* Alias for **toCSSUnclamped**

**toHexString** *()* `'#ff00ff'` CSS hex, clamped with no alpha. 

**toRGBA** *()* `[1,0,1,1]` Raw values.

**toObject** *()* `{ r: 1, g: 0, b: 1, a: 1}` Raw values.

**toBuffer** *()* `Buffer.from([255,0,255,255])` Clamped, 8-bit values.

**to8BitArrays** *( length )* `[255,0,255,255]` Clamped, 8-bit values.

**toHexChannels** *( 'rgbahsv' )* `'ff00ffd4ffff'` Clamped, 8-bit values from RGB and HSV as hex.

**toCSS** *( format = 'auto' | 'rgba' | 'hex' | 'unclamped' )* CSS in given format.

**toCSSRGBA** *()* `'rgb(255,0,255)'` Clamped, 8-bit values with auto adding alpha.

**toCSSUnclamped** *()* `'rgba(1024,0,1024,1)'` Unclamped values in CSS range, always has alpha.

## Other Methods

* **isBlack** *() -> bool* `red == 0 && green == 0 && blue == 0`
* **isGray** *() -> bool* `red == green == blue`


# Hue Preservation

In [HSV & HLS](https://en.wikipedia.org/wiki/HSL_and_HSV) colour spaces, hue and saturation can be invalid when the colour is grey or black. In purely functional colour conversions, this results in `saturation` being lost when `value == 0` and `hue` being lost when `saturation == 0`. This is particularly annoying in stateful system, where `value` or `saturation` are being modulated.

`deepcolour` compensates for this by holding `hue` and `saturation` values.

``` js
// Initialize a new Colour
let colour = new Colour('green')
assert.equal( colour.hue, 1/3)

// Set the colour to black
colour.value = 0
assert( colour.isBlack() )

// Since it does not matter when value is 0,
// hue is preserved.
assert.equal( colour.hue, 1/3)

// We can set the hue, as well.
colour.hue = 1/6

// Colour is still black
assert( colour.isBlack() )

// But the hue is preserved
assert.equal( colour.hue, 1/6)

// Set the colour back to full intensity
colour.value = 1

// Is now yellow
assert.equal( colour.hex, '#ffff00' )
```


# Vectors

By default `deepcolour` creates a colour space with `rgba` properties. The library can also be used to create and manipulate arbitrary vector spaces, with or without colour properties.

## Colourless Vectors

``` js
const { Space } = require('deepcolour')
const Vector = Space({
  rgba: false,
  channels: 'xyz'
})

let vec = Vector.add( [ 1, 0, 0 ], { y: 0.5 } )
assert.deepEqual( vec.toArray(), [ 1, 0.5, 0 ] )
```

## Full Vertex

``` js
const { Space } = require('deepcolour')
const Vertex = Space({
  rgba: 5,
  channels: 'xyzuvrgba'
})

let vert = new Vertex( { x: 3, y: 2, z: 1 } )

// All colour manipulation methods may be used.
vert.set('cyan')
vert.alpha = 0.5

assert.deepEqual( vert.toArray(), [ 3, 2, 1, 0, 0, 0, 1, 1, 0.5 ] )
```

CHANGELOG.md
# Changelog

## [2.2.0] - 2020-03-03

### Added
- Support for hue / saturation preservation when setting from RGB.
- This file.

### Changed
- Misc reshuffling.
- Tests to check additions.

src/Colour.js
'use strict'

const NUM_CHANNELS = 4
const COLOUR_NAMES = require('color-name')

class Colour {
  constructor() {
    this.setDefault()
    this.setArguments( arguments )
  }

  clone() {
    const clone = new Colour()

    this.eachChannel( function ( v, c ) {
      clone[c] = v
    } )

    return clone
  }

  eachChannel( callback ) {
    for ( var c = 0; c < NUM_CHANNELS; c++ ) {
      var result = callback.call( this, this[c], c )
      result = parseFloat( result )
      if ( !isNaN( result ) )
        this[c] = result
    }
  }

  channelByName( name ) {
    switch ( name ) {
      case '0': case 'r': case 'red':
        return this.red

      case '1': case 'g': case 'green':
        return this.green

      case '2': case 'b': case 'blue':
        return this.blue

      case '3': case 'a': case 'alpha':
        return this.alpha

      case 'h': case 'hue':
        return this.hue

      case 's': case 'sat': case 'saturation':
        return this.saturation

      case 'v': case 'value':
        return this.value
    }

    return NaN
  }

  channel8Bit( channel ) {
    var value = this[channel]
    value = parseFloat( value )
    value = value < 0 ? 0 : value > 1 ? 1 : value
    value = Math.round( value * 255 )
    return value
  }

  isRGBNormal() {
    return this[0] >= 0
        && this[0] <= 1
        && this[1] >= 0
        && this[1] <= 1
        && this[2] >= 0
        && this[2] <= 1
  }

  isNormal() {
    return this[0] >= 0
        && this[0] <= 1
        && this[1] >= 0
        && this[1] <= 1
        && this[2] >= 0
        && this[2] <= 1
        && this[3] >= 0
        && this[3] <= 1
  }

  isGray() {
    return this[0] == this[1]
        && this[1] == this[2]
  }

  isBlack() {
    return this[0] == 0
        && this[1] == 0
        && this[2] == 0
  }

  //
  // RGB getters and setters
  //
  set red   ( value ) {
    value = parseCSSValue( value )
    if ( !isNaN( value ) )
      this[0] = value
  }
  set green ( value ) {
    value = parseCSSValue( value )
    if ( !isNaN( value ) )
      this[1] = value
  }
  set blue  ( value ) {
    value = parseCSSValue( value )
    if ( !isNaN( value ) )
      this[2] = value
  }
  set alpha ( value ) {
    value = parseCSSAlpha( value )
    if ( !isNaN( value ) )
      this[3] = value
  }

  get red   () { return this[0] }
  get green () { return this[1] }
  get blue  () { return this[2] }
  get alpha () { return this[3] }

  //
  // HSV getters and setters
  //
  set hue ( value ) {
    value = parseCSSHue( value )
    this.setHSV( value, this.saturation, this.value  )
  }

  set saturation ( value ) {
    value = parseCSSValue( value )
    this.setHSV( this.hue, value, this.value  )
  }

  set value ( value ) {
    value = parseCSSValue( value )
    this.setHSV( this.hue, this.saturation, value  )
  }

  get value () {
    return Math.max( this[0], this[1], this[2] )
  }

  get saturation () {
    if ( this.isBlack() )
      return this._saturation || 0

    const max = Math.max( this[0], this[1], this[2] )
        , min = Math.min( this[0], this[1], this[2] )
        , delta = max - min

    return max > 0 ? delta / max : 0
  }

  get hue() {
    const max = Math.max( this[0], this[1], this[2] )
        , min = Math.min( this[0], this[1], this[2] )
        , delta = max - min

    if ( delta <= 0 )
      return this._hue || 0

    const red = this[0]
        , green = this[1]
        , blue = this[2]

    var hue
    if ( red == max )
      hue = ( green - blue ) / delta
    else if ( green == max )
      hue = 2 + ( blue - red ) / delta
    else
      hue = 4 + ( red - green ) / delta

    if ( hue < 0 )
      hue += 6

    hue /= 6

    // Hack to allow hue >= 1 to be returned when
    // set specifically by setHSV
    if ( this._hue && ( this._hue % 1 ) == hue )
      hue = this._hue

    this._hue = hue

    return hue
  }

  //
  // String getters / setters
  //

  get css() {
    return this.toCSS()
  }

  set css( val ) {
    this.setString( val )
  }

  get hex() {
    return this.toHexString()
  }

  set hex( val ) {
    this.setString( val )
  }

  //
  // Chainable setStuff functions
  //

  set( ob ) {

    if ( ob === null || ob == undefined )
      return this

    if ( Colour.isColour( ob ) || Array.isArray( ob ) ) {
      this.eachChannel( ( v, c ) => ob[c] )

      return this
    }

    if ( 'string' == typeof ob )
      return this.setString( ob )

    if ( global.Buffer && Buffer.isBuffer( ob ) )
      return this.set8BitArray( ob )

    if ( ob !== null ) {
      return this.setKeys( ob )
    }

    return this
  }


  setDefault() {
    this[0] = 0
    this[1] = 0
    this[2] = 0
    this[3] = 1

    return this
  }

  setArguments( args, fill = false ) {
    var argChannel = 0
    for ( var argI = 0; argI < args.length; argI++ ) {
      var arg = args[argI]
      if ( 'number' == typeof arg ) {
        this.setChannel( argChannel++, arg )
      } else if ( 'string' == typeof arg ) {
        this.setString( arg )
        argChannel = 3
      } else if ( 'undefined' !== typeof arg ) {
        this.set( arg )
      }
    }

    while( fill && argChannel && argChannel < NUM_CHANNELS ) {
      this.setChannel( argChannel, this[argChannel-1] )
      argChannel ++
    }

    return this
  }

  setKeys( ob ) {
    const self = this
        , props = [
            ['red','r'],
            ['green','g'],
            ['blue','b'],
            ['alpha','a'],
            ['hue','h'],
            ['saturation','sat','s'],
            ['value','v'],
            ['hex'],
            ['css']
          ]

    if ( 'object' == typeof ob )
      props.forEach( function( keys ) {
        var prop
        keys.forEach( function ( key ) {
          prop = prop || key
          if ( key in ob )
            self[prop] = ob[key]
        } )
      } )

    return this
  }

  setAlpha( alpha ) {
    alpha = parseFloat( alpha )
    if ( !isNaN( alpha ) ) this[3] = alpha
    return this
  }

  setRandom() {
    const R = Math.random
    return this.setRGB( R(), R(), R() )
  }

  setRGB( r, g, b, a ) {
    const args = arguments.length == 1 && arguments[0].length ? arguments[0] : arguments

    r = parseCSSValue( args[0] )
    g = parseCSSValue( args[1] )
    b = parseCSSValue( args[2] )
    a = parseCSSAlpha( args[3] )

    if ( !isNaN( r ) ) this[0] = r
    if ( !isNaN( g ) ) this[1] = g
    if ( !isNaN( b ) ) this[2] = b
    if ( !isNaN( a ) ) this[3] = a

    return this
  }

  setHSV( hue, sat, val ) {
    hue = parseFloat( hue )
    sat = parseFloat( sat )
    val = parseFloat( val )

    if ( isNaN( hue ) ) hue = this.hue
    if ( isNaN( sat ) ) sat = this.saturation
    if ( isNaN( val ) ) val = this.value

    this._hue = hue
    hue = hue % 1
    hue = hue < 0 ? hue + 1 : hue
    this._saturation = sat


    if ( sat == 0 ) {
      this.setRGB( val, val, val )
      return this
    }

    const hex = Math.floor( hue * 6 )
        , inHex = hue * 6 - hex
        , eP = val * (1 - sat)
        , eQ = val * (1 - (sat * inHex))
        , eT = val * (1 - (sat * (1 - inHex)))

    switch ( hex ) {
      case 0: this.setRGB( val, eT, eP ); break
      case 1: this.setRGB( eQ, val, eP ); break
      case 2: this.setRGB( eP, val, eT ); break
      case 3: this.setRGB( eP, eQ, val ); break
      case 4: this.setRGB( eT, eP, val ); break
      case 5: this.setRGB( val, eP, eQ ); break
    }

    return this
  }

  set8BitArray( input ) {
    if ( !input || 'number' != typeof input.length )
      throw new Error('Expected input to be array-like.')

    this.eachChannel( ( value, c ) => input[c] / 255 )
    return this
  }

  setChannel( channel, value ) {
    channel = parseInt( channel ) || 0
    if ( channel < 0 || channel > 3 )
      throw new Error('Invalid channel')

    value = parseFloat( value )
    if ( !isNaN( value ) )
      this[channel] = value

    return this
  }

  setChannelHex( channel, value ) {
    if ( 'string' == typeof value ) {
      if ( !value.length || value.length > 2 )
        throw new Error('Invalid hex value input' )

      if ( value.length < 2 )
        value = value + value
  
      value = parseInt( value, 16 )      
    } 

    if ( 'number' != typeof value ) 
      throw new Error('Invalid input' )

    value /= 255

    return this.setChannel( channel, value )
  }

  setChannelByName( name, value ) {
    if ( 'string' !== typeof name )
      throw new Error('channel name must be string')

    switch ( name.toLowerCase() ) {
      case 'r': case 'red':
        this.red = value
      break

      case 'g': case 'green':
        this.green = value
      break

      case 'b': case 'blue':
        this.blue = value
      break

      case 'a': case 'alpha':
        this.alpha = value
      break

      case 'h': case 'hue':
        this.hue = value
      break

      case 's': case 'sat': case 'saturation':
        this.saturation = value
      break

      case 'v': case 'value':
        this.value = value
      break
    }

    return this
  }

  setString( str ) {
    if ( 'transparent' == str )
      return this.setAlpha( 0 )

    str = str.toLowerCase()

    if ( str in COLOUR_NAMES )
      return this.set8BitArray( COLOUR_NAMES[str] )

    let match

    if ( match = /hsla?\(\s*(-?[\d\.]+%?)\s*,\s*(-?[\d\.]+%)\s*,\s*(-?[\d\.]+%)\s*(,\s*(-?[\d\.]+%?))?\s*\)/.exec( str ) ) {
      this.setHSL( match[1], match[2], match[3], match[5] )
    } else if ( match = /rgba?\(\s*(-?[\d\.]+%?)\s*,\s*(-?[\d\.]+%?)\s*,\s*(-?[\d\.]+%?)\s*(,\s*(-?[\d\.]+%?))?\s*\)/.exec( str ) ) {
        this.setRGB(
          match[1],
          match[2],
          match[3],
          match[5]
        )
    } else if ( match = /^#?([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})/.exec( str ) ) {
      this.setChannelHex( 0, match[1] )
      this.setChannelHex( 1, match[2] )
      this.setChannelHex( 2, match[3] )
    } else if ( match = /^#?([0-9a-f])([0-9a-f])([0-9a-f])/.exec( str ) ) {
      this.setChannelHex( 0, match[1] )
      this.setChannelHex( 1, match[2] )
      this.setChannelHex( 2, match[3] )
    }

    return this
  }

  setHSL() {
    var args = arguments.length == 1 && arguments[0].length ? arguments[0] : arguments
    const h = parseCSSHue( args[0] )
        , s = parseCSSValue( args[1] )
        , l = parseCSSValue( args[2] )
        , alpha = parseCSSAlpha( args[3] )


    // Shamelessly ganked from http://stackoverflow.com/a/9493060
    var r, g, b;
    if(s == 0){
        r = g = b = l; // achromatic
    } else {
        var hue2rgb = function hue2rgb(p, q, t){
            if(t < 0) t += 1;
            if(t > 1) t -= 1;
            if(t < 1/6) return p + (q - p) * 6 * t;
            if(t < 1/2) return q;
            if(t < 2/3) return p + (q - p) * (2/3 - t) * 6;
            return p;
        }

        var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
        var p = 2 * l - q;
        r = hue2rgb(p, q, h + 1/3);
        g = hue2rgb(p, q, h);
        b = hue2rgb(p, q, h - 1/3);
    }

    return this.setRGB( r, g, b, alpha )
  }

  toString() {
    return this.toCSSUnclamped()
  }

  valueOf() {
    return this.toCSSUnclamped()
  }

  inspect() {
    return this.toString()
  }

  //
  //
  //

  toHexString() {
    var result = '#'
    for ( var c = 0; c < 3; c ++ )
      result += valueToHex( this[c] )

    return result
  }

  toRGBA() {
    return [
      this[0],
      this[1],
      this[2],
      this[3]
    ]
  }

  toObject( keys ) {
    if ( !keys )
      keys = [ 'r','g','b','a' ]

    if ( !Array.isArray( keys ) )
      throw new Error('keys must be array')

    const result = {}

    keys.forEach( ( key ) => {
      result[key] = this.channelByName( key )
    })

    return result
  }

  toBuffer( length ) {
    length = parseInt( length )
    if ( isNaN( length ) )
      length = 4

    if ( length < 0 || length > 4 )
      throw new Exception( 'Invalid length' )

    const buffer = Buffer.alloc( length )

    for ( var c = 0; c < length; c++ )
      buffer.writeUInt8( this.channel8Bit( c ), c )

    return buffer
  }

  to8BitArray( length ) {
    length = parseInt( length )
    if ( isNaN( length ) )
      length = 4

    if ( length < 0 || length > 4 )
      throw new Exception( 'Invalid length' )

    const array = new Array( length )
    for ( var c = 0; c < length; c++ )
      array[c] = this.channel8Bit( c )

    return array
  }

  toHexChannels( channels ) {
    let result = ''

    channels = channels || 'rgb'
    if ( 'string' != typeof channels )
      throw new Error('channels must be string')

    for ( let i = 0; i < channels.length; i ++ ) {
      let value = this.channelByName( channels[i] )
      value = value || 0
      value = valueToHex( value )
      result += value
    }

    return result
  }

  toCSS( format = 'auto' ) {
    switch( format ) {
      case 'rgba':
        return this.toCSSRGBA()
        
      case 'hex':
        return this.toHexString()

      case 'unclamped':
        return this.toCSSUnclamped()

      default:
      case 'auto':
        let alpha = this.alpha
        alpha = clampValue( alpha )
        if ( this.isRGBNormal() && alpha == 1 ) 
          return this.toHexString()

        return this.toCSSRGBA()
    }
  }
  toCSSRGBA() {
    let numbers = this.to8BitArray(3)
      , tag = 'rgb'
      , alpha = this.alpha

    alpha = clampValue( alpha )

    if ( alpha != 1 ) {
      numbers.push( valueToAlpha( alpha ) )
      tag += 'a'
    }

    let inner = numbers.join( ',' )

    return `${tag}(${inner})`
  }

  toCSSUnclamped() {
    let numbers = this.toRGBA()
    numbers = numbers.map( ( value, channel ) => 
      channel < 3 ? 
        Math.floor( value * 255 ).toString() :
        valueToAlpha( value )
    )
    let inner = numbers.join( ',' )

    return `rgba(${inner})`    
  }


  //
  // Composite operators
  //
  mix( b, amount ) {
    amount = parseAmount( amount )
    b = parseColour( b )
    this.eachChannel( ( value, channel ) =>
      value * ( 1 - amount ) + b[channel] * amount
    )
  }

  add() {
    let b = new Colour()
    b.setArguments( arguments )
    this.eachChannel( ( value, channel ) => value + b[channel] )
    return this
  }

  multiply() {
    let b = new Colour()
    b.setArguments( arguments, true )
    this.eachChannel( ( value, channel ) => value * b[channel] )
    return this
  }


}

Colour.isColour = function ( object ) {
  return object instanceof Colour
}

Colour.equal = function() {
  const colour = new Colour()
  var str
  for ( var i = 0; i < arguments.length; i ++ ) {
    colour.set( arguments[i] )
    if ( !str )
      str = colour.toString()
    else if ( str != colour.toString() )
      return false
  }

  return true
}

module.exports = Colour

//
// Utility functions
//

function clampValue( a ) {
  a = parseFloat( a ) || 0
  if ( a > 1 )
    return 1
  
  if ( a < 0 )
    return 0

  return a
}

function parseCSSHue( v ) {
  if ( 'number' == typeof v ) return v
  if ( !v )  return NaN

  return parseFloat( v ) / 360
}

function parseCSSValue( v ) {
  if ( 'number' == typeof v ) return v
  if ( !v )  return NaN

  var div = 255
  if ( v.substr( v.length - 1 ) == '%' )
    div = 100

  return parseFloat( v ) / div
}

function parseCSSAlpha( v ) {
  if ( 'number' == typeof v ) return v
  if ( !v )  return NaN

  var div = 1
  if ( v.substr( v.length - 1 ) == '%' )
    div = 100

  return parseFloat( v ) / div
}

function valueToHex( v ) {
  v = v < 0 ? 0 : v > 1 ? 1 : v
  v = Math.round( v * 255 )
  v = v.toString( 16 )
  if ( v.length < 2 )
    v = '0'+v

  return v
}

function valueToPercent( v ) {
  v *= 100
  v = v.toFixed( 1 )
  v += '%'

  return v
}

function valueToAlpha( v ) {
  return v.toFixed( 2 )
}

function parseColour( colour ) {
  if ( !(colour instanceof Colour ) )
    throw new Error(`Ain't a Colour`)

  return colour
}

function parseAmount( amount ) {
  amount = parseFloat( amount )
  if ( isNaN( amount ) )
    return 1

  return amount
}
